src/vm.rs

```diff
@@ -414,34 +454,75 @@ impl<'a, S: Storage, C: PevmChain> Database for VmDb<'a, S, C> {
     }
 
     fn storage(&mut self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        if let LazyStrategy::ERC20Transfer {
+            sender_balance_slot,
+            recipient_balance_slot,
+        } = self.lazy_strategy
+        {
+            if index == sender_balance_slot {
+                return self
+                    .vm
+                    .storage
+                    .storage(&address, &index)
+                    .map_err(|err| ReadError::StorageError(err.to_string()));
+            }
+            if index == recipient_balance_slot {
+                return self
+                    .vm
+                    .storage
+                    .storage(&address, &index)
+                    .map_err(|err| ReadError::StorageError(err.to_string()));
+            }
+        }
+
         let location_hash = self
             .vm
             .hasher
             .hash_one(MemoryLocation::Storage(address, index));
 
         let read_origins = self.read_set.entry(location_hash).or_default();
+        read_origins.clear();
```

Comment:
Let's at least add a `TODO` to proper handle read origins to avoid race conditions.

Reply:
Actually I think calling `push_origin` is pretty straightforward and can be done right away?

Reply:
I have copy-pasted the logic from `basic`. No more race conditions.

The current handling is quite messy though.